design/DESIGN.md

```diff
@@ -0,0 +1,94 @@
+
+**Simplifying the State Machines**
+
+One option would be to attempt to simplify Brupop’s current state machine such that the structure of the machine is clear by reading data in the source code, and deriving operation of the system from the declarative description. While this would make it simpler to add features to the existing system, we would still rely on annotations and labels to propagate information.
+
+### Prometheus for Visualization
+
+The [Prometheus](https://prometheus.io/) project is a [CNCF-graduated](https://github.com/cncf/toc/blob/main/process/graduation_criteria.adoc) monitoring solution for cloud-native applications. Prometheus’ status in CNCF and popularity with Cloud-native customers means that Kubernetes cluster operators will likely be familiar with operating  Prometheus in their clusters, and will be familiar with how to monitor the Prometheus endpoint exposed by Brupop. Indeed, [customers asking for improved observability](https://github.com/bottlerocket-os/bottlerocket-update-operator/issues/57) have mentioned Prometheus integration directly. Alternatives to Prometheus include:
+
+* Integration with an AWS Service like CloudWatch
+    * Unfavorable, as we’d like to support Kubernetes clusters outside of AWS just as much as those within.
+* Expose logs to a service that we spin up within the customer’s cluster, like Kibana.
+    * Spinning up our own visualization frontend is a waste of our customers’ resources, especially if they are already using their own visualization frontend, like Prometheus.
+* Exposing logs through a custom API from the Brupop controller.
+    * Offering our own metrics API for customers to scrape means they must build their own components to handle those metrics and ingest them into their own systems. This seems like a strictly worse option than Prometheus integration
+
+## Appendix
+
+### Appendix A. State Transitions, Error Cases
+
+Many things can go wrong while scheduling updates. In some cases, poor node health may indicate the need to wait til later. If a problem seems to be indicative of a botched update, we will want to cease operation of the update agent until an operator can resolve the issue. Further still, it may be possible via use of the Kubernetes `drain` API to enter a stuck state (see [“stuck evictions”](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#stuck-evictions) and [examples in the PodDisruptionBudget documentation](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pdb-example)).
```

Comment:
```suggestion
Many things can go wrong while scheduling updates. In some cases, poor node health may indicate the need to wait till later. If a problem seems to be indicative of a botched update, we will want to cease operation of the update agent until an operator can resolve the issue. Further still, it may be possible via use of the Kubernetes `drain` API to enter a stuck state (see [“stuck evictions”](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#stuck-evictions) and [examples in the PodDisruptionBudget documentation](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pdb-example)).
```